libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+#if defined(GPIO_CAN2_RX)
+        stm32_configgpio(GPIO_CAN2_RX | GPIO_PULLUP);
+        stm32_configgpio(GPIO_CAN2_TX);
+#endif
+    }
+#if !defined(GPIO_CAN1_RX) &&  !defined(GPIO_CAN2_RX)
+# error  "Need to define GPIO_CAN1_RX and/or GPIO_CAN2_RX"
+#endif
+
+    /*
+     * IRQ
+     */
+    if (can_number >= 1) {
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN1TX, can1_irq);
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN1RX0, can1_irq);
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN1RX1, can1_irq);
+    }
+
+    if (can_number >= 1) {
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN2TX, can2_irq);
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN2RX0, can2_irq);
+        CAN_IRQ_ATTACH(STM32_IRQ_CAN2RX1, can2_irq);
+    }
+}
+
+int PX4CANDriver::init(const uint32_t bitrate, const PX4CAN::OperatingMode mode, uint8_t can_number)
+{
+    int res = 0;
+
+    can_number_ = can_number;
+
+    if (AP_BoardConfig::get_can_debug() >= 2) {
+        printf("PX4CANDriver::init Bitrate %lu mode %d\n\r", static_cast<unsigned long>(bitrate), static_cast<int>(mode));
+    }
+
+    static bool initialized_once = false;
+    if (!initialized_once) {
+        initialized_once = true;
+        if (AP_BoardConfig::get_can_debug() >= 2) {
+            printf("PX4CANDriver::init First initialization\n\r");
+        }
+        initOnce(can_number);
+    }
+
+    /*
+     * CAN1
+     */
+    if (can_number >= 1) {
+        if (AP_BoardConfig::get_can_debug() >= 2) {
+            printf("PX4CANDriver::init Initing iface 0...\n\r");
+        }
+        ifaces[0] = &if0_;               // This link must be initialized first,
+        res = if0_.init(bitrate, mode); // otherwise an IRQ may fire while the interface is not linked yet;
+        if (res < 0) {             // a typical race condition.
+            if (AP_BoardConfig::get_can_debug() >= 2) {
+                printf("PX4CANDriver::init Iface 0 init failed %i\n\r", res);
+            }
+            ifaces[0] = UAVCAN_NULLPTR;
+        }
+    }
+
+    /*
+     * CAN2
+     */
+    if (can_number >= 2) {
+#if UAVCAN_STM32_NUM_IFACES > 1
+        if (AP_BoardConfig::get_can_debug() >= 2) {
+            printf("PX4CANDriver::init Initing iface 1...\n\r");
+        }
+        ifaces[1] = &if1_;                          // Same thing here.
+        res = if1_.init(bitrate, mode);
+        if (res < 0) {
+            if (AP_BoardConfig::get_can_debug() >= 2) {
+                printf("PX4CANDriver::init Iface 1 init failed %i\n\r", res);
+            }
+            ifaces[1] = UAVCAN_NULLPTR;
+        }
+#endif
+    }
+
+    if (AP_BoardConfig::get_can_debug() >= 2) {
+        printf("PX4CANDriver::init CAN drv init OK, res = %d\n\r", res);
+    }
+
+    return res;
+}
+
+PX4CAN* PX4CANDriver::getIface(uint8_t iface_index)
+{
+    if (iface_index < UAVCAN_STM32_NUM_IFACES) {
+        return ifaces[iface_index];
+    }
+    return UAVCAN_NULLPTR;
+}
+
+bool PX4CANDriver::hadActivity()
+{
+    bool ret = if0_.hadActivity();
+#if UAVCAN_STM32_NUM_IFACES > 1
```

Comment:
Ok, but there are several places where we have can_number available and we still use the macro too. I think that can_number should be checked on entry points to be less or equal to the macro and changed to the macro if it isn't; then we can always rely on can_number having the correct number of usable interfaces.